File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/geth/tracer.ex
defmodule EthereumJSONRPC.Geth.Tracer do
  @moduledoc """
  Elixir implementation of a custom tracer (`priv/js/ethereum_jsonrpc/geth/debug_traceTransaction/tracer.js`)
  for variants that don't support specifying tracer in [debug_traceTransaction](https://github.com/ethereum/go-ethereum/wiki/Management-APIs#debug_tracetransaction) calls.
  """

  import EthereumJSONRPC, only: [integer_to_quantity: 1, quantity_to_integer: 1]

  def replay(%{"structLogs" => logs} = result, receipt, tx) when is_list(logs) do
    %{"contractAddress" => contract_address} = receipt
    %{"from" => from, "to" => to, "value" => value, "input" => input} = tx

    top =
      to
      |> if do
        %{
          "type" => "call",
          "callType" => "call",
          "to" => to,
          "input" => input,
          "output" => Map.get(result, "return", "0x" <> Map.get(result, "returnValue", ""))
        }
      else
        %{
          "type" => "create",
          "init" => input,
          "createdContractAddressHash" => contract_address,
          "createdContractCode" => "0x"
        }
      end
      |> Map.merge(%{
        "from" => from,
        "traceAddress" => [],
        "value" => value,
        "gas" => 0,
        "gasUsed" => 0
      })

    ctx = %{
      depth: 1,
      stack: [top],
      trace_address: [0],
      calls: [[]]
    }

    logs
    |> Enum.reduce(ctx, &step/2)
    |> finalize()
  end

  defp step(%{"error" => _}, %{stack: [%{"error" => _} | _]} = ctx), do: ctx

  defp step(
         %{"error" => _} = log,
         %{
           depth: stack_depth,
           stack: [call | stack],
           trace_address: [_, trace_index | trace_address],
           calls: [subsubcalls, subcalls | calls]
         } = ctx
       ) do
    call = process_return(log, Map.put(call, "error", "error"))

    subsubcalls =
      subsubcalls
      |> Enum.reverse()
      |> Enum.map(fn
        subcalls when is_list(subcalls) -> subcalls
        subcall when is_map(subcall) -> %{subcall | "from" => call["createdContractAddressHash"] || call["to"]}
      end)

    %{
      ctx
      | depth: stack_depth - 1,
        stack: stack,
        trace_address: [trace_index + 1 | trace_address],
        calls: [[subsubcalls, call | subcalls] | calls]
    }
  end

  defp step(
         %{"depth" => log_depth} = log,
         %{
           depth: stack_depth,
           stack: [call | stack],
           trace_address: [_, trace_index | trace_address],
           calls: [subsubcalls, subcalls | calls]
         } = ctx
       )
       when log_depth == stack_depth - 1 do
    call = process_return(log, call)

    subsubcalls =
      subsubcalls
      |> Enum.reverse()
      |> Enum.map(fn
        subcalls when is_list(subcalls) -> subcalls
        subcall when is_map(subcall) -> %{subcall | "from" => call["createdContractAddressHash"] || call["to"]}
      end)

    step(log, %{
      ctx
      | depth: stack_depth - 1,
        stack: stack,
        trace_address: [trace_index + 1 | trace_address],
        calls: [[subsubcalls, call | subcalls] | calls]
    })
  end

  defp step(%{"gas" => log_gas, "gasCost" => log_gas_cost} = log, %{stack: [%{"gas" => call_gas} = call | stack]} = ctx) do
    gas = max(call_gas, log_gas)
    op(log, %{ctx | stack: [%{call | "gas" => gas, "gasUsed" => gas - log_gas - log_gas_cost} | stack]})
  end

  defp op(%{"op" => "CREATE"} = log, ctx), do: create_op(log, ctx)
  defp op(%{"op" => "CREATE2"} = log, ctx), do: create_op(log, ctx, "create2")
  defp op(%{"op" => "SELFDESTRUCT"} = log, ctx), do: self_destruct_op(log, ctx)
  defp op(%{"op" => "CALL"} = log, ctx), do: call_op(log, "call", ctx)
  defp op(%{"op" => "CALLCODE"} = log, ctx), do: call_op(log, "callcode", ctx)
  defp op(%{"op" => "DELEGATECALL"} = log, ctx), do: call_op(log, "delegatecall", ctx)
  defp op(%{"op" => "STATICCALL"} = log, ctx), do: call_op(log, "staticcall", ctx)
  defp op(%{"op" => "REVERT"}, ctx), do: revert_op(ctx)
  defp op(_, ctx), do: ctx

  defp process_return(%{"stack" => log_stack}, %{"type" => "create"} = call) do
    [ret | _] = Enum.reverse(log_stack)

    case quantity_to_integer(ret) do
      0 -> Map.put(call, "error", call["error"] || "internal failure")
      _ -> %{call | "createdContractAddressHash" => "0x" <> String.slice(ret, 24, 40)}
    end
  end

  defp process_return(
         %{"stack" => log_stack, "memory" => log_memory},
         %{"outputOffset" => out_off, "outputLength" => out_len} = call
       ) do
    [ret | _] = Enum.reverse(log_stack)

    ret
    |> quantity_to_integer()
    |> case do
      0 ->
        Map.put(call, "error", call["error"] || "internal failure")

      _ ->
        output =
          log_memory
          |> IO.iodata_to_binary()
          |> String.slice(out_off, out_len)

        %{call | "output" => "0x" <> output}
    end
    |> Map.drop(["outputOffset", "outputLength"])
  end

  defp create_op(
         %{"stack" => log_stack, "memory" => log_memory},
         %{depth: stack_depth, stack: stack, trace_address: trace_address, calls: calls} = ctx,
         type \\ "create"
       ) do
    [value, input_offset, input_length | _] = Enum.reverse(log_stack)

    init =
      log_memory
      |> IO.iodata_to_binary()
      |> String.slice(quantity_to_integer("0x" <> input_offset) * 2, quantity_to_integer("0x" <> input_length) * 2)

    call = %{
      "type" => type,
      "from" => nil,
      "traceAddress" => Enum.reverse(trace_address),
      "init" => "0x" <> init,
      "gas" => 0,
      "gasUsed" => 0,
      "value" => "0x" <> value,
      "createdContractAddressHash" => nil,
      "createdContractCode" => "0x"
    }

    %{
      ctx
      | depth: stack_depth + 1,
        stack: [call | stack],
        trace_address: [0 | trace_address],
        calls: [[] | calls]
    }
  end

  defp self_destruct_op(
         %{"stack" => log_stack, "gas" => log_gas, "gasCost" => log_gas_cost},
         %{trace_address: [trace_index | trace_address], calls: [subcalls | calls]} = ctx
       ) do
    [to | _] = Enum.reverse(log_stack)

    if quantity_to_integer(to) in 1..8 do
      ctx
    else
      call = %{
        "type" => "selfdestruct",
        "from" => nil,
        "to" => "0x" <> String.slice(to, 24, 40),
        "traceAddress" => Enum.reverse([trace_index | trace_address]),
        "gas" => log_gas,
        "gasUsed" => log_gas_cost,
        "value" => "0x0"
      }

      %{ctx | trace_address: [trace_index + 1 | trace_address], calls: [[call | subcalls] | calls]}
    end
  end

  defp call_op(
         %{"stack" => log_stack, "memory" => log_memory},
         call_type,
         %{
           depth: stack_depth,
           stack: [%{"value" => parent_value} = parent | stack],
           trace_address: trace_address,
           calls: calls
         } = ctx
       ) do
    [_, to | log_stack] = Enum.reverse(log_stack)

    {value, [input_offset, input_length, output_offset, output_length | _]} =
      case call_type do
        "delegatecall" ->
          {parent_value, log_stack}

        "staticcall" ->
          {"0x0", log_stack}

        _ ->
          [value | rest] = log_stack
          {"0x" <> value, rest}
      end

    input =
      log_memory
      |> IO.iodata_to_binary()
      |> String.slice(quantity_to_integer("0x" <> input_offset) * 2, quantity_to_integer("0x" <> input_length) * 2)

    call = %{
      "type" => "call",
      "callType" => call_type,
      "from" => nil,
      "to" => "0x" <> String.slice(to, 24, 40),
      "traceAddress" => Enum.reverse(trace_address),
      "input" => "0x" <> input,
      "output" => "0x",
      "outputOffset" => quantity_to_integer("0x" <> output_offset) * 2,
      "outputLength" => quantity_to_integer("0x" <> output_length) * 2,
      "gas" => 0,
      "gasUsed" => 0,
      "value" => value
    }

    %{
      ctx
      | depth: stack_depth + 1,
        stack: [call, parent | stack],
        trace_address: [0 | trace_address],
        calls: [[] | calls]
    }
  end

  defp revert_op(%{stack: [last | stack]} = ctx) do
    %{ctx | stack: [Map.put(last, "error", "execution reverted") | stack]}
  end

  defp finalize(%{stack: [top], calls: [calls]}) do
    calls =
      Enum.map(calls, fn
        subcalls when is_list(subcalls) -> subcalls
        subcall when is_map(subcall) -> %{subcall | "from" => top["createdContractAddressHash"] || top["to"]}
      end)

    [top | Enum.reverse(calls)]
    |> List.flatten()
    |> Enum.map(fn %{"gas" => gas, "gasUsed" => gas_used} = call ->
      %{call | "gas" => integer_to_quantity(gas), "gasUsed" => integer_to_quantity(gas_used)}
    end)
  end
end

File: apps/ethereum_jsonrpc/priv/js/ethereum_jsonrpc/geth/debug_traceTransaction/tracer.js
// tracer allows Geth's `debug_traceTransaction` to mimic the output of Parity's `trace_replayTransaction`
{
    // The call stack of the EVM execution.
    callStack: [{}],

    // step is invoked for every opcode that the VM executes.
    step(log, db) {
        // Capture any errors immediately
        const error = log.getError();

        if (error !== undefined) {
            this.fault(log, db);
        } else {
            this.success(log, db);
        }
    },

    // fault is invoked when the actual execution of an opcode fails.
    fault(log, db) {
        // If the topmost call already reverted, don't handle the additional fault again
        if (this.topCall().error === undefined) {
            this.putError(log);
        }
    },

    putError(log) {
        if (this.callStack.length > 1) {
            this.putErrorInTopCall(log);
        } else {
            this.putErrorInBottomCall(log);
        }
    },

    putErrorInTopCall(log) {
        // Pop off the just failed call
        const call = this.callStack.pop();
        this.putErrorInCall(log, call);
        this.pushChildCall(call);
    },

    putErrorInBottomCall(log) {
        const call = this.bottomCall();
        this.putErrorInCall(log, call);
    },

    putErrorInCall(log, call) {
        call.error = log.getError();

        // Consume all available gas and clean any leftovers
        if (call.gasBigInt !== undefined) {
            call.gasUsedBigInt = call.gasBigInt;
        }

        delete call.outputOffset;
        delete call.outputLength;
    },

    topCall() {
        return this.callStack[this.callStack.length - 1];
    },

    bottomCall() {
        return this.callStack[0];
    },

    pushChildCall(childCall) {
        const topCall = this.topCall();

        if (topCall.calls === undefined) {
            topCall.calls = [];
        }

        topCall.calls.push(childCall);
    },

    pushGasToTopCall(log) {
        const topCall = this.topCall();

        if (topCall.gasBigInt === undefined) {
            topCall.gasBigInt = log.getGas();
        }
        topCall.gasUsedBigInt = topCall.gasBigInt - log.getGas() - log.getCost();
    },

    success(log, db) {
        const op = log.op.toString();

        this.beforeOp(log, db);

        switch (op) {
            case 'CREATE':
                this.createOp(log);
                break;
            case 'CREATE2':
                this.create2Op(log);
                break;
            case 'SELFDESTRUCT':
                this.selfDestructOp(log, db);
                break;
            case 'CALL':
            case 'CALLCODE':
            case 'DELEGATECALL':
            case 'STATICCALL':
                this.callOp(log, op);
                break;
            case 'REVERT':
                this.revertOp();
                break;
        }
    },

    beforeOp(log, db) {
        /**
         * Depths
         * 0 - `ctx`.  Never shows up in `log.getDepth()`
         * 1 - first level of `log.getDepth()`
         *
         * callStack indexes
         *
         * 0 - pseudo-call stand-in for `ctx` in initializer (`callStack: [{}]`)
         * 1 - first callOp inside of `ctx`
         */
        const logDepth = log.getDepth();
        const callStackDepth = this.callStack.length;

        if (logDepth < callStackDepth) {
            // Pop off the last call and get the execution results
            const call = this.callStack.pop();

            const ret = log.stack.peek(0);

            if (!ret.equals(0)) {
                if (call.type === 'create' || call.type === 'create2') {
                    call.createdContractAddressHash = toHex(toAddress(ret.toString(16)));
                    call.createdContractCode = toHex(db.getCode(toAddress(ret.toString(16))));
                } else {
                    call.output = toHex(log.memory.slice(call.outputOffset, call.outputOffset + call.outputLength));
                }
            } else if (call.error === undefined) {
                call.error = 'internal failure';
            }

            delete call.outputOffset;
            delete call.outputLength;

            this.pushChildCall(call);
        }
        else {
            this.pushGasToTopCall(log);
        }
    },

    createOp(log) {
        const inputOffset = log.stack.peek(1).valueOf();
        const inputLength = log.stack.peek(2).valueOf();
        const inputEnd = inputOffset + inputLength;
        const stackValue = log.stack.peek(0);

        const call = {
            type: 'create',
            from: toHex(log.contract.getAddress()),
            init: toHex(log.memory.slice(inputOffset, inputEnd)),
            valueBigInt: bigInt(stackValue.toString(10))
        };
        this.callStack.push(call);
    },

    create2Op(log) {
        const inputOffset = log.stack.peek(1).valueOf();
        const inputLength = log.stack.peek(2).valueOf();
        const inputEnd = inputOffset + inputLength;
        const stackValue = log.stack.peek(0);

        const call = {
            type: 'create2',
            from: toHex(log.contract.getAddress()),
            init: toHex(log.memory.slice(inputOffset, inputEnd)),
            valueBigInt: bigInt(stackValue.toString(10))
        };
        this.callStack.push(call);
    },

    selfDestructOp(log, db) {
        const contractAddress = log.contract.getAddress();

        this.pushChildCall({
            type: 'selfdestruct',
            from: toHex(contractAddress),
            to: toHex(toAddress(log.stack.peek(0).toString(16))),
            gasBigInt: log.getGas(),
            gasUsedBigInt: log.getCost(),
            valueBigInt: db.getBalance(contractAddress)
        });
    },

    callOp(log, op) {
        const to = toAddress(log.stack.peek(1).toString(16));

        // Skip any pre-compile invocations, those are just fancy opcodes
        if (!isPrecompiled(to)) {
            this.callCustomOp(log, op, to);
        }
    },

    callCustomOp(log, op, to) {
        const stackOffset = (op === 'DELEGATECALL' || op === 'STATICCALL' ? 0 : 1);

        const inputOffset = log.stack.peek(2 + stackOffset).valueOf();
        const inputLength = log.stack.peek(3 + stackOffset).valueOf();
        const inputEnd = inputOffset + inputLength;

        const call = {
            type: 'call',
            callType: op.toLowerCase(),
            from: toHex(log.contract.getAddress()),
            to: toHex(to),
            input: toHex(log.memory.slice(inputOffset, inputEnd)),
            outputOffset: log.stack.peek(4 + stackOffset).valueOf(),
            outputLength: log.stack.peek(5 + stackOffset).valueOf()
        };

        switch (op) {
            case 'CALL':
            case 'CALLCODE':
                call.valueBigInt = bigInt(log.stack.peek(2));
                break;
            case 'DELEGATECALL':
                // value inherited from scope during call sequencing
                break;
            case 'STATICCALL':
                // by definition static calls transfer no value
                call.valueBigInt = bigInt.zero;
                break;
            default:
                throw 'Unknown custom call op ' + op;
        }

        this.callStack.push(call);
    },

    revertOp() {
        this.topCall().error = 'execution reverted';
    },

    // result is invoked when all the opcodes have been iterated over and returns
    // the final result of the tracing.
    result(ctx, db) {
        const result = this.ctxToResult(ctx, db);
        const filtered = this.filterNotUndefined(result);
        const callSequence = this.sequence(filtered, [], filtered.valueBigInt, []).callSequence;
        return this.encodeCallSequence(callSequence);
    },

    ctxToResult(ctx, db) {
        var result;

        switch (ctx.type) {
            case 'CALL':
                result = this.ctxToCall(ctx);
                break;
            case 'CREATE':
                result = this.ctxToCreate(ctx, db);
                break;
            case 'CREATE2':
                result = this.ctxToCreate2(ctx, db);
                break;
        }

        return result;
    },

    ctxToCall(ctx) {
        const result = {
            type: 'call',
            callType: 'call',
            from: toHex(ctx.from),
            to: toHex(ctx.to),
            valueBigInt: bigInt(ctx.value.toString(10)),
            gasBigInt: bigInt(ctx.gas),
            gasUsedBigInt: bigInt(ctx.gasUsed),
            input: toHex(ctx.input)
        };

        this.putBottomChildCalls(result);
        this.putErrorOrOutput(result, ctx);

        return result;
    },

    putErrorOrOutput(result, ctx) {
        const error = this.error(ctx);

        if (error !== undefined) {
            result.error = error;
        } else {
            result.output = toHex(ctx.output);
        }
    },

    ctxToCreate(ctx, db) {
        const result = {
            type: 'create',
            from: toHex(ctx.from),
            init: toHex(ctx.input),
            valueBigInt: bigInt(ctx.value.toString(10)),
            gasBigInt: bigInt(ctx.gas),
            gasUsedBigInt: bigInt(ctx.gasUsed)
        };

        this.putBottomChildCalls(result);
        this.putErrorOrCreatedContract(result, ctx, db);

        return result;
    },

    ctxToCreate2(ctx, db) {
        const result = {
            type: 'create2',
            from: toHex(ctx.from),
            init: toHex(ctx.input),
            valueBigInt: bigInt(ctx.value.toString(10)),
            gasBigInt: bigInt(ctx.gas),
            gasUsedBigInt: bigInt(ctx.gasUsed)
        };

        this.putBottomChildCalls(result);
        this.putErrorOrCreatedContract(result, ctx, db);

        return result;
    },

    putBottomChildCalls(result) {
        const bottomCall = this.bottomCall();
        const bottomChildCalls = bottomCall.calls;

        if (bottomChildCalls !== undefined) {
            result.calls = bottomChildCalls;
        }
    },

    putErrorOrCreatedContract(result, ctx, db) {
        const error = this.error(ctx);

        if (error !== undefined) {
            result.error = error
        } else {
            result.createdContractAddressHash = toHex(ctx.to);
            if (toHex(ctx.input) != '0x') {
              result.createdContractCode = toHex(db.getCode(ctx.to));
            } else {
              result.createdContractCode = '0x';
            }
        }
    },

    error(ctx) {
        var error;

        const bottomCall = this.bottomCall();
        const bottomCallError = bottomCall.error;

        if (bottomCallError !== undefined) {
            error = bottomCallError;
        } else {
            const ctxError = ctx.error;

            if (ctxError !== undefined) {
                error = ctxError;
            }
        }

        return error;
    },

    filterNotUndefined(call) {
        for (var key in call) {
            if (call[key] === undefined) {
                delete call[key];
            }
        }

        if (call.calls !== undefined) {
            for (var i = 0; i < call.calls.length; i++) {
                call.calls[i] = this.filterNotUndefined(call.calls[i]);
            }
        }

        return call;
    },

    // sequence converts the finalized calls from a call tree to a call sequence
    sequence(call, callSequence, availableValueBigInt, traceAddress) {
        const subcalls = call.calls;
        delete call.calls;

        call.traceAddress = traceAddress;

        if (call.type === 'call' && call.callType === 'delegatecall') {
            call.valueBigInt = availableValueBigInt;
        }

        var newCallSequence = callSequence.concat([call]);

        if (subcalls !== undefined) {
            for (var i = 0; i < subcalls.length; i++) {
                const nestedSequenced = this.sequence(
                    subcalls[i],
                    newCallSequence,
                    call.valueBigInt,
                    traceAddress.concat([i])
                );
                newCallSequence = nestedSequenced.callSequence;
            }
        }

        return {
            callSequence: newCallSequence
        };
    },

    encodeCallSequence(calls) {
        for (var i = 0; i < calls.length; i++) {
            this.encodeCall(calls[i]);
        }

        return calls;
    },

    encodeCall(call) {
        this.putValue(call);
        this.putGas(call);
        this.putGasUsed(call);

        return call;
    },

    putValue(call) {
        const valueBigInt = call.valueBigInt;
        delete call.valueBigInt;

        call.value = '0x' + valueBigInt.toString(16);
    },

    putGas(call) {
        let gasBigInt = call.gasBigInt;
        delete call.gasBigInt;

        if (gasBigInt === undefined) {
            gasBigInt = bigInt.zero;
        }

        call.gas = '0x' + gasBigInt.toString(16);
    },

    putGasUsed(call) {
        let gasUsedBigInt = call.gasUsedBigInt;
        delete call.gasUsedBigInt;

        if (gasUsedBigInt === undefined) {
            gasUsedBigInt = bigInt.zero;
        }

        call.gasUsed = '0x' + gasUsedBigInt.toString(16);
    }
}
